### PR TITLE
[move-2024] Generalize name access parsing to handle macros, tyargs

### PR DESCRIPTION
**Pull request by cgswords**
## Description 

This generalizes name access chains in parsing to handle tyargs, toward allowing `Option<Int>::Some` in enums. Macros weren't part of the goal, but since macros are `foo::bar!<args>(...)`, it became necessary.

## Test Plan 

A few new tests, plus everything works (with a few shifted expected outputs, hopefully for the better).

---
If your changes are not user-facing and do not break anything, you can skip the following section. Otherwise, please briefly describe what has changed under the Release Notes section.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/expansion/path_expander.rs [hunk 1]
```diff
@@ -0,0 +1,1105 @@
...
+                        }
+                        Some(mident) => {
+                            let n2_name = n2.name;
+                            let (tyargs, is_macro) = if !(path.has_tyargs_last()) {
+                                let mut diag = diag!(
+                                    Syntax::InvalidName,
+                                    (path.tyargs_loc().unwrap(), "Invalid type argument position")
+                                );
+                                diag.add_note(
+                                    "Diagnostics must appear at the end of a name access path",
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/expansion/path_expander.rs:953]: Highlighting again here, I'm not sure what diagnostics means in this context

File: external-crates/move/crates/move-compiler/src/expansion/path_expander.rs [hunk 2]
```diff
@@ -0,0 +1,1105 @@
...
+                            top_level_address(context, /* suggest_declaration */ false, *ln);
+                        let mident = sp(ident_loc, ModuleIdent_::new(addr, ModuleName(n2.name)));
+                        let access = EN::ModuleAccess(mident, n3.name);
+                        let (tyargs, is_macro) = if !(path.has_tyargs_last()) {
+                            let mut diag = diag!(
+                                Syntax::InvalidName,
+                                (path.tyargs_loc().unwrap(), "Invalid type argument position")
+                            );
+                            diag.add_note(
+                                "Diagnostics must appear at the end of a name access path",
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/expansion/path_expander.rs:983]: ditto

File: external-crates/move/crates/move-compiler/src/expansion/path_expander.rs [hunk 3]
```diff
@@ -0,0 +1,1105 @@
...
+                    _ if path.entries.len() > 2 => {
+                        let mut diag = diag!(Syntax::InvalidName, (loc, "Too many name segments"));
+                        diag.add_note("Names may only have 0, 1, or 2 segments separated by '::'");
+                        context.env.add_diag(diag);
+                        return None;
+                    }
+                    _ => {
+                        let diag = diag!(
+                            Syntax::InvalidName,
+                            (loc, "This name has an unexpected format")
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/expansion/path_expander.rs:1001]: Is this possible to trigger?

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/match_keyword.move [hunk 4]
```diff
@@ -1,3 +0,0 @@
-module a::m {
-    fun match(_: u64): bool { false }
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/match_keyword.move:2]: Why was this removed?

**Review by tnowacki — COMMENTED** [review]
Still have to review the parser, but some initial comments

File: external-crates/move/crates/move-compiler/src/parser/ast.rs [hunk 5]
```diff
@@ -301,16 +306,65 @@ pub struct Constant {
...
+    Single(PathEntry),
+    Path(NamePath),
+    // // <Name>
+    // One(Name),
+    // // (<Name>|<Num>)::<Name>, bool indicates last name was pushed on
+    // Two(LeadingNameAccess, Name, bool),
+    // // (<Name>|<Num>)::<Name>::<Name>, bool indicates last name was pushed on
+    // Three(Spanned<(LeadingNameAccess, Name)>, Name, bool),
+    // // (<Name>|<Num>)::<Name>::<Name>::<Name>, bool indicates last name was pushed on
+    // Four(Spanned<(LeadingNameAccess, Name)>, Name, Name, bool),
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/ast.rs:367]: nit, removing is probably fine at this point

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/parser/ast.rs:367]: Ditto for a bunch below

File: external-crates/move/crates/move-compiler/src/parser/ast.rs [hunk 6]
```diff
@@ -301,16 +306,65 @@ pub struct Constant {
...
+}
+
+// A path root.
+// For now these should hever have tyargs or macro call set (though the type arguments will be
+// used for enums).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RootPathEntry {
+    pub name: LeadingNameAccess,
+    pub tyargs: Option<Spanned<Vec<Type>>>,
+    pub is_macro: Option<Loc>,
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/ast.rs:344]: I didn't see any tests for this? I figure based on conversations that it was a later edition, but would still be good to test

File: external-crates/move/crates/move-compiler/src/parser/ast.rs [hunk 7]
```diff
@@ -301,16 +306,65 @@ pub struct Constant {
...
+//       | Leading :: Name :: Name <TyArgs>
+//       | Leading :: Name :: Name <TyArgs> :: Name
+//
+// VName = MName <TyArgs> :: Name
+
 // A ModuleAccess references a local or global name or something from a module,
 // either a struct type or a function.
+// The boolean flag indicates if a name was pushed on after tyarg parsing, which we use in
+// expansion to determine if the variant form was valid.
+// FIXME(cswords): We should eventually rewrite all of this to be `NamePath`s that hold tyargs:
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/ast.rs:322]: I don't quite understand the FIXME, in what you were going for below

File: external-crates/move/crates/move-compiler/src/parser/ast.rs [hunk 8]
```diff
@@ -650,6 +711,253 @@ impl LeadingNameAccess_ {
...
+    fn is_macro(&self) -> Option<&Loc> {
+        if let Some(loc) = &self.root.is_macro {
+            return Some(loc);
+        }
+        for entry in self.entries.iter() {
+            if let Some(loc) = &entry.is_macro {
+                return Some(loc);
+            }
+        }
+        None
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/ast.rs:811]: Not sure if its really better but....
```suggestion
        self.root.is_macro.as_ref().or_else(|| self.entries.iter().find_map(|e| e.is_macro.as_ref()))
```

File: external-crates/move/crates/move-compiler/src/parser/ast.rs [hunk 9]
```diff
@@ -650,6 +711,253 @@ impl LeadingNameAccess_ {
...
+    fn len(&self) -> usize {
+        1 + self.entries.len()
+    }
+}
+
+macro_rules! forward_name_access {
+    ($self:ident.$call:ident($($args:ident),*)) => {
+        match $self {
+            NameAccessChain_::Single(entry) => entry.$call($($args),*),
+            NameAccessChain_::Path(entry) => entry.$call($($args),*),
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/ast.rs:921]: It might not be called enum_dispatch... but I know what it is ;)

Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/parser/ast.rs:921]: Honestly I probably prefer this lol

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs [hunk 10]
```diff
@@ -336,6 +340,29 @@ where
     }
 }
 
+// Helper for location blocks
+
+macro_rules! ok_with_loc {
+    ($context:expr, $body:block) => {{
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/src/parser/syntax.rs:346]: Are the blocks not expressions?

File: external-crates/move/crates/move-compiler/src/parser/ast.rs (hunk 6 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/parser/ast.rs:344]: Added a test.

File: external-crates/move/crates/move-compiler/src/parser/syntax.rs (hunk 10 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/parser/syntax.rs:346]: I have had rust complain before that they are not, but it does appear to matter situation-dependent and works just fine here

File: external-crates/move/crates/move-compiler/src/expansion/path_expander.rs (hunk 3 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/expansion/path_expander.rs]: No, it should be an `ice!` and is now.

File: external-crates/move/crates/move-compiler/src/parser/ast.rs (hunk 7 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/parser/ast.rs]: Old comment from before I did the work "properly" -- I was going to do the ad-hoc solution we suggested. Sorry for the confusion.

File: external-crates/move/crates/move-compiler/tests/move_check/parser/acquires_list_generic.exp [hunk 11]
```diff
@@ -1,9 +1,6 @@
-error[E01002]: unexpected token
+error[E01016]: invalid name
   ┌─ tests/move_check/parser/acquires_list_generic.move:6:25
   │
 6 │     fun foo() acquires B<CupC<R>> {
-  │                         ^
-  │                         │
-  │                         Unexpected '<'
-  │                         Expected ','
+  │                         ^^^^^^^^^ Type arguments are disallowed here. Expected a resource struct name
```
[the remark this answers is not in the document]
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_check/parser/acquires_list_generic.exp]: Done and done!

File: external-crates/move/crates/move-compiler/src/parser/ast.rs (hunk 9 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/src/parser/ast.rs:879]: Yeah, I think I like this more, too. The overall thing is clear without too much "magic."

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move [hunk 12]
```diff
@@ -0,0 +1,73 @@
+module a::m {
+
+    public struct S<T> { u: T }
+
+    public macro fun make_s<$T>($u: T): S<T> {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move:5]:
```suggestion
    public macro fun make_s<$T>($u: $T): S<$T> {
```

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        S{ u: $u }
+    }
+}
+
+module a::m {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move:10]:
```suggestion
module a::n {
```

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        a!::m!::make_s!<u64>(0u64)
+    }
+
+}
+
+module 0x42::m {
+
+    public struct S<T> { u: T }
+
+    public macro fun make_s<$T>($u: T): S<T> {
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move:42]:
```suggestion
    public macro fun make_s<$T>($u: $T): S<$T> {
```

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        S{ u: $u }
+    }
+}
+
+module a::m {
+
+    fun valid(): a::m::S<u64> {
+        a::m::make_s!<u64>(0u64)
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move:13]: Nit, I try to avoid mixing valid and invalid examples in tests. Personally, I find it hard to to then grok the `.exp` file

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    fun test00(): a::m::S<u64> {
+        a::m::make_s!<u64>(0u64)
```
Comment (tnowacki) [inline external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move:17]: Is this intended to be the same as `valid`

**Review by tnowacki — APPROVED** [review]
Approved with nits/testing issues

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move (hunk 15 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move]: Removed

File: external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move (hunk 16 above)
Comment (cgswords) [reply external-crates/move/crates/move-compiler/tests/move_2024/parser/invalid_macro_locs.move]: Removed
